### PR TITLE
Stream Forwarding Changes.

### LINKED ISSUE #925: Standard output redirection hides current line
The way stdout (and stderr) is redirected from chained commands prevents the last line from being show until after an end-of-line character. This causes weird interactions.

```C#
Console.WriteLine();
Console.Write("Press any key to continue . . . ");
Console.ReadKey(intercept: true);
```

With a tool containing the above code, invoking it (e.g. via `dotnet mytool`) will...

1. Print a blank line
2. Wait for a keypress
3. Print `Press any key to continue . . . `

The expected behavior would be...

1. Print a blank line
3. Print `Press any key to continue . . . `
2. Wait for a keypress

### PR DESCRIPTION
**Pull request by brthor**
Stream Forwarding changes to not wait on buffer full before writing. Instead child process input streams will be read character by character as they Console.Write or Console.WriteLine. Upon finding a newline character, the line will be printed to the parent process's console.

On child process exit, any remaining characters will be flushed.

This will partially fix #925 with the caveat that a newline must be written to the child process stream before any output will be written. When the newline is written, order of writes is now preserved though.

Previously, any write which did not fill the Streamforwarder's internal buffer would be held until the buffer was full or the process exited.

PTAL @anurse @piotrpMSFT @Sridhar-MS @davidfowl @livarcocc

### CODE REVIEW

File: src/Microsoft.DotNet.Cli.Utils/StreamForwarder.cs [hunk 1]
```diff
@@ -7,45 +7,45 @@ namespace Microsoft.DotNet.Cli.Utils
...
 
-        public void ForwardTo(Action<string> write, Action<string> writeLine)
+        public StreamForwarder ForwardTo(Action<string> writeLine)
         {
             if (writeLine == null)
             {
                 throw new ArgumentNullException(nameof(writeLine));
             }
+
             if (_writeLine != null)
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/StreamForwarder.cs]: Can you encapsulate this in a method? Like WriteLineAlreadySet? I was asking myself how this was going to work. I had to read this like twice to notice the _.

**Comment by livarcocc** [comment]
LGTM

File: src/Microsoft.DotNet.Cli.Utils/StreamForwarder.cs (hunk 1 above)
Comment (brthor) [inline src/Microsoft.DotNet.Cli.Utils/StreamForwarder.cs]: Yeah I had the same first reaction. I'll go ahead and make this change.

**Comment by brthor** [comment]
cc @mmitche 
Seeing this in Windows

ERROR: [WS-CLEANUP] Cannot delete workspace: remote file operation failed: d:\j\workspace\dotnet_cli\release_windows_nt_prtest at hudson.remoting.Channel@5694d2f2:dci-win-tbld-29: java.nio.file.AccessDeniedException: d:\j\workspace\dotnet_cli\release_windows_nt_prtest\artifacts\tests\coreclr.dll

Aware of this issue? Going to retest so here's a link: 
http://dotnet-ci.cloudapp.net/job/dotnet_cli/job/release_windows_nt_prtest/1329/

**Comment by brthor** [comment]
@dotnet-bot test Windows_NT Release Build Please

**Comment by mmitche** [comment]
@brthor Thanks for the heads up.  I think these are process that were hung but orphaned when Jenkins had a restart.

**Comment by brthor** [comment]
@dotnet-bot test Windows_NT Debug Build please

**Comment by brthor** [comment]
@dotnet-bot test this please

**Comment by brthor** [comment]
@dotnet-bot test OSX Debug Build please

**Comment by mmitche** [comment]
@brthor It looks like something changed with the test reports.  Multiple machines have showed anability to convert the reports.  Any ideas?

**Comment by mmitche** [comment]
@brthor The regex for the hung process/file issue is picking up the xunit conversion failure.  Please take a look at the results logs before retesting.

**Comment by brthor** [comment]
@cdmihai Looks like I am getting a build test failure.

```
15:04:14     Microsoft.DotNet.Tools.Builder.Tests.IncrementalTests.TestNoLockFile [FAIL]
15:04:14       Assert.Contains() Failure
15:04:14       Not found: does not have a lock file
```

Would you mind taking a quick look?
http://dotnet-ci.cloudapp.net/job/dotnet_cli/job/debug_windows_nt_prtest/1397/console

**Comment by cdmihai** [comment]
@brthor That test asserts that build prints an error message when it is called on a project that has no lock file. I did not have time to look over the PR diff. Do you skip restoring the test project?

**Comment by brthor** [comment]
@cdmihai Ah Thanks! I imagine this might be because StdErr got redirected back to StdErr on the console.

**Comment by brthor** [comment]
@pakrym Build Output tests are failing on CentOS. Looks like you have a casing issue.

All other tests are good. going to merge this! :smile:

**Pull request merged**